api/src/main/java/com/flipkart/poseidon/ds/trie/Trie.java

```diff
@@ -37,58 +36,89 @@ public Trie() {
         TrieNode currentParent = root;
-        for (int i = 0; i < keys.length; i++) {
-            TrieNode matchingNode = findMatchingNode(currentNode, keys[i]);
+        for (int i = 0; i < keys.size(); i++) {
+            TrieNode matchingNode = findMatchingNode(currentNode, keys.get(i));
             if (matchingNode != null) {
                 currentParent = matchingNode;
                 currentNode = matchingNode.firstChild;
-                matchingNode.value = i == keys.length - 1 ? value : matchingNode.value;
+                matchingNode.value = i == keys.size() - 1 ? value : matchingNode.value;
                 continue;
             }
 
             TrieNode newNode = new TrieNode();
-            newNode.key = keys[i];
-            newNode.matchAny = keys[i] == null;
-            newNode.value = i == keys.length - 1 ? value : null;
+            newNode.key = keys.get(i).key;
+            newNode.matchAny = keys.get(i).wildCard;
+            newNode.greedyMatchAny = keys.get(i).greedyWildCard;
+            newNode.value = i == keys.size() - 1 ? value : null;
 
             if (currentNode == null) {
                 currentParent.firstChild = newNode;
-            } else if (newNode.key != null) {
+            } else if (!newNode.matchAny && !newNode.greedyMatchAny) {
                 newNode.rightSibling = currentParent.firstChild;
                 currentParent.firstChild = newNode;
-            } else {
-                while (currentNode.rightSibling != null) {
-                    currentNode = currentNode.rightSibling;
+            } else if (newNode.matchAny) {
+                if (currentNode.greedyMatchAny) {
+                    newNode.rightSibling = currentNode;
+                    currentParent.firstChild = newNode;
+                } else {
+                    while (currentNode.rightSibling != null) {
+                        if (currentNode.rightSibling.greedyMatchAny) {
+                            break;
+                        }
+                        currentNode = currentNode.rightSibling;
+                    }
+
+                    if (currentNode.rightSibling != null && currentNode.rightSibling.greedyMatchAny) {
+                        newNode.rightSibling = currentNode.rightSibling;
+                        currentNode.rightSibling = newNode;
+                    } else {
+                        currentNode.rightSibling = newNode;
+                    }
+
+                    currentParent.wildChild = newNode;
+                }
+            } else if (newNode.greedyMatchAny) {
+                if (currentNode.matchAny) {
+                    currentParent.rightSibling = newNode;
+                } else {
+                    while (currentNode.rightSibling != null) {
+                        currentNode = currentNode.rightSibling;
+                    }
+
+                    currentNode.rightSibling = newNode;
+                    currentParent.greedyWildChild = newNode;
```

Comment:
We'll detach concrete nodes, wild card and greedy wild card. Then the last two else if should simplify a lot. Please change get() accordingly